### DISCUSSION TITLE
launch application reinstall the application instead of launching it. Why ?

### DISCUSSION THREAD

**Discussion by efrenette11** [appium/appium#16641] [Q&A]
PLease help.

No metter what I try, I can reopen the application during a session without reinstalling it. Any idea ?

**Comment by jlipps** [comment]
What have you tried already?

**Comment by efrenette11** [comment]
I tried try play with noReset and full reset values without success.

**Comment by efrenette11** [comment]
I'm using robot framework

**Comment by efrenette11** [comment]
Here's a pretty simple case I wrote to demo this.

As you can see in the test setup, I'll launch the application but this will remove it install or simply reopen it. At the end, I also added the logs that I get from appium when launch command wa executed. I would like to understand why "Installing '/Users/efrenette/Downloads/gotoconnect.app'" happens ?

*** Keywords ***
Prepare Suite
    &{invocationArgs}=  Create dictionary
                         ...  platformName=iOS
                         ...  platformVersion=14.5
                         ...  deviceName=iPhone 11
                         ...  app=/Users/efrenette/Downloads/gotoconnect.app
    ${dump}=  Get Dictionary Items  ${invocationArgs}
    Log   \nOpening application with those capabilities : ${dump}    console=yes
    AppiumLibrary.open application
                         ...  http://localhost:4723/wd/hub
                         ...  &{invocationArgs}
    Quit application

Test Setup
    Launch application
    Generate Random Contact Details



Logs


The server is running
[HTTP] --> POST /wd/hub/session/14e07689-80c8-41fa-9274-85068f50360e/appium/app/launch
[HTTP] {}
[W3C (14e07689)] Calling AppiumDriver.launchApp() with args: ["14e07689-80c8-41fa-9274-85068f50360e"]
[XCUITest] Executing command 'launchApp'
[XCUITest] Current user: 'efrenette'
[XCUITest] Available devices: 
[XCUITest] No real device with udid 'DE5E37AE-12F1-4FF5-808A-0D759EEEEA60'. Looking for simulator
[iOSSim] Constructing iOS simulator for Xcode version 12.5.1 with udid 'DE5E37AE-12F1-4FF5-808A-0D759EEEEA60'
[XCUITest] Determining device to run tests on: udid: 'DE5E37AE-12F1-4FF5-808A-0D759EEEEA60', real device: false
[BaseDriver] Event 'xcodeDetailsRetrieved' logged at 1648063854104 (15:30:54 GMT-0400 (Eastern Daylight Time))
[BaseDriver] Using local app '/Users/efrenette/Downloads/gotoconnect.app'
[BaseDriver] Event 'appConfigured' logged at 1648063854104 (15:30:54 GMT-0400 (Eastern Daylight Time))
[XCUITest] Checking whether app '/Users/efrenette/Downloads/gotoconnect.app' is actually present on file system
[XCUITest] App is present
[BaseDriver] Event 'resetStarted' logged at 1648063854105 (15:30:54 GMT-0400 (Eastern Daylight Time))
[XCUITest] Not scrubbing third party app in anticipation of uninstall
[BaseDriver] Event 'resetComplete' logged at 1648063854902 (15:30:54 GMT-0400 (Eastern Daylight Time))
[WebDriverAgent] Using WDA path: '/Applications/Appium.app/Contents/Resources/app/node_modules/appium/node_modules/appium-webdriveragent'
[WebDriverAgent] Using WDA agent: '/Applications/Appium.app/Contents/Resources/app/node_modules/appium/node_modules/appium-webdriveragent/WebDriverAgent.xcodeproj'
[XCUITest] Trying to start iOS log capture but it has already started!
[BaseDriver] Event 'logCaptureStarted' logged at 1648063854912 (15:30:54 GMT-0400 (Eastern Daylight Time))
[XCUITest] Setting up simulator
[iOS] No reason to set locale
[iOS] No iOS / app preferences to set
[iOS] Setting did not need to be updated
[iOSSim] Setting preferences of DE5E37AE-12F1-4FF5-808A-0D759EEEEA60 Simulator to {"ConnectHardwareKeyboard":false}
[iOSSim] Setting common Simulator preferences to {"RotateWindowWhenSignaledByGuest":true,"ConnectHardwareKeyboard":false,"PasteboardAutomaticSync":false}
[iOSSim] Updated DE5E37AE-12F1-4FF5-808A-0D759EEEEA60 Simulator preferences at '/Users/efrenette/Library/Preferences/com.apple.iphonesimulator.plist' with {"RotateWindowWhenSignaledByGuest":true,"ConnectHardwareKeyboard":false,"PasteboardAutomaticSync":false,"DevicePreferences":{"DE5E37AE-12F1-4FF5-808A-0D759EEEEA60":{"ConnectHardwareKeyboard":false,"SimulatorExternalDisplay":2114,"SimulatorWindowRotationAngle":0,"SimulatorWindowGeometry":{"A16E1889-8DC8-4182-94E5-43C77C0464DE":{"WindowCenter":"{388, 577}","WindowScale":1},"3082E2E1-CF61-4427-9F2F-CCD302A0C992":{"WindowCenter":"{1385, 579}","WindowScale":1},"AB51DF45-8226-4620-AAFE-CF8C3773F8A3":{"WindowCenter":"{310, 584}","WindowScale":1},"87A2BC8A-E005-462F-A4A2-7E55B5623A0A":{"WindowCenter":"{481, 508}","WindowScale":1}},"SimulatorWindowOrientation":"Portrait","ChromeTint":""}}}
[iOSSim] Got Simulator UI client PID: 550
[iOSSim] Both Simulator with UDID 'DE5E37AE-12F1-4FF5-808A-0D759EEEEA60' and the UI client are currently running
[BaseDriver] Event 'simStarted' logged at 1648063855669 (15:30:55 GMT-0400 (Eastern Daylight Time))
[XCUITest] Verifying application platform
[XCUITest] CFBundleSupportedPlatforms: ["iPhoneSimulator"]
[XCUITest] Reset requested. Removing app with id 'com.logmein.gotoconnect' from the device
[XCUITest] Installing '/Users/efrenette/Downloads/gotoconnect.app' on Simulator with UUID 'DE5E37AE-12F1-4FF5-808A-0D759EEEEA60'...
[WebDriverAgent] Parsed BUILD_DIR configuration value: '/Users/efrenette/Library/Developer/Xcode/DerivedData/WebDriverAgent-alwvnomvwrdtzoaxbbkniqrpcdpp/Build/Products'
[WebDriverAgent] Got derived data root: '/Users/efrenette/Library/Developer/Xcode/DerivedData/WebDriverAgent-alwvnomvwrdtzoaxbbkniqrpcdpp'
[XCUITest] The app has been installed successfully.
[BaseDriver] Event 'appInstalled' logged at 1648063860268 (15:31:00 GMT-0400 (Eastern Daylight Time))
[WebDriverAgent] No obsolete cached processes from previous WDA sessions listening on port 8100 have been found
[DevCon Factory] Requesting connection for device DE5E37AE-12F1-4FF5-808A-0D759EEEEA60 on local port 8100
[DevCon Factory] Cached connections count: 0
[DevCon Factory] Successfully requested the connection for DE5E37AE-12F1-4FF5-808A-0D759EEEEA60:8100
[XCUITest] Starting WebDriverAgent initialization with the synchronization key 'XCUITestDriver'
[WD Proxy] Matched '/status' to command name 'getStatus'
[WD Proxy] Proxying [GET /status] to [GET http://127.0.0.1:8100/status] with no body
[WD Proxy] Got response with status 200: {"value":{"message":"WebDriverAgent is ready to accept commands","state":"success","os":{"testmanagerdVersion":28,"name":"iOS","sdkVersion":"14.5","version":"14.5"},"ios":{"simulatorVersion":"14.5","ip":"192.168.1.62"},"ready":true,"build":{"upgradedAt":"1632042036000","time":"Mar 23 2022 13:23:55","productBundleIdentifier":"com.facebook.WebDriverAgentRunner"}},"sessionId":null}
[WebDriverAgent] Upgrade timestamp of the currently bundled WDA: 1632042036000
[WebDriverAgent] Upgrade timestamp of the WDA on the device: 1632042036000
[WebDriverAgent] Will reuse previously cached WDA instance at 'http://127.0.0.1:8100/' with 'com.facebook.WebDriverAgentRunner'. Set the wdaLocalPort capability to a value different from 8100 if this is an undesired behavior.
[XCUITest] Trying to start WebDriverAgent 2 times with 10000ms interval
[XCUITest] These values can be customized by changing wdaStartupRetries/wdaStartupRetryInterval capabilities
[BaseDriver] Event 'wdaStartAttempted' logged at 1648063860369 (15:31:00 GMT-0400 (Eastern Daylight Time))
[WebDriverAgent] Using provided WebdriverAgent at 'http://127.0.0.1:8100/'
[WD Proxy] Matched '/status' to command name 'getStatus'
[WD Proxy] Proxying [GET /status] to [GET http://127.0.0.1:8100/status] with no body
[WD Proxy] Got response with status 200: {"value":{"message":"WebDriverAgent is ready to accept commands","state":"success","os":{"testmanagerdVersion":28,"name":"iOS","sdkVersion":"14.5","version":"14.5"},"ios":{"simulatorVersion":"14.5","ip":"192.168.1.62"},"ready":true,"build":{"upgradedAt":"1632042036000","time":"Mar 23 2022 13:23:55","productBundleIdentifier":"com.facebook.WebDriverAgentRunner"}},"sessionId":null}
[BaseDriver] Event 'wdaSessionAttempted' logged at 1648063860373 (15:31:00 GMT-0400 (Eastern Daylight Time))
[XCUITest] Sending createSession command to WDA
[WD Proxy] Matched '/session' to command name 'createSession'
[WD Proxy] Proxying [POST /session] to [POST http://127.0.0.1:8100/session] with body: {"capabilities":{"firstMatch":[{"bundleId":"com.logmein.gotoconnect","arguments":[],"environment":{},"eventloopIdleDelaySec":0,"shouldWaitForQuiescence":true,"shouldUseTestManagerForVisibilityDetection":false,"maxTypingFrequency":60,"shouldUseSingletonTestManager":true,"shouldTerminateApp":true}],"alwaysMatch":{}}}
[WD Proxy] Got response with status 200: {"value":{"sessionId":"4D906F33-C6DD-4C5A-90FB-001A2A222B68","capabilities":{"device":"iphone","browserName":"GoTo","sdkVersion":"14.5","CFBundleIdentifier":"com.logmein.gotoconnect"}},"sessionId":"4D906F33-C6DD-4C5A-90FB-001A2A222B68"}
[WD Proxy] Determined the downstream protocol as 'W3C'
[BaseDriver] Event 'wdaSessionStarted' logged at 1648063862465 (15:31:02 GMT-0400 (Eastern Daylight Time))
[BaseDriver] Event 'wdaStarted' logged at 1648063862465 (15:31:02 GMT-0400 (Eastern Daylight Time))
[XCUITest] Skipping setting of the initial display orientation. Set the "orientation" capability to either "LANDSCAPE" or "PORTRAIT", if this is an undesired behavior.
[BaseDriver] Event 'orientationSet' logged at 1648063862465 (15:31:02 GMT-0400 (Eastern Daylight Time))
[iOS] Successfully launched the '/Users/efrenette/Downloads/gotoconnect.app' app.
[W3C (14e07689)] Responding to client with driver.launchApp() result: null
[HTTP] <-- POST /wd/hub/session/14e07689-80c8-41fa-9274-85068f50360e/appium/app/launch 200 8789 ms - 14
[HTTP] 
[BaseDriver] Shutting down because we waited 60 seconds for a command
[Appium] Closing session, cause was 'New Command Timeout of 60 seconds expired. Try customizing the timeout using the 'newCommandTimeout' desired capability'
[Appium] Removing session 'cf9f14b3-765a-41cd-a5ed-ae6d303ada28' from our master session list
[DevCon Factory] Releasing connections for DE5E37AE-12F1-4FF5-808A-0D759EEEEA60 device on any port number
[DevCon Factory] Found cached connections to release: ["DE5E37AE-12F1-4FF5-808A-0D759EEEEA60:8100"]
[DevCon Factory] Cached connections count: 0
[XCUITest] Not clearing log files. Use `clearSystemFiles` capability to turn on.
[BaseDriver] Shutting down because we waited 60 seconds for a command
[Appium] Closing session, cause was 'New Command Timeout of 60 seconds expired. Try customizing the timeout using the 'newCommandTimeout' desired capability'
[Appium] Removing session '14e07689-80c8-41fa-9274-85068f50360e' from our master session list
[DevCon Factory] Releasing connections for DE5E37AE-12F1-4FF5-808A-0D759EEEEA60 device on any port number
[DevCon Factory] No cached connections have been found
[XCUITest] Not clearing log files. Use `clearSystemFiles` capability to turn on.
[IOSSimulatorLog] Stopping iOS log capture

**Comment by efrenette11** [comment]
Whit noReset=true I was expecting the app to not be removed

**Comment by efrenette11** [comment]
Here's another try I had 

Prepare Suite
    Run Process  xcrun simctl keychain booted reset  shell=yes
    &{invocationArgs}=  Create dictionary
                         ...  platformName=iOS
                         ...  platformVersion=14.5
                         ...  deviceName=iPhone 11
                         ...  app=/Users/efrenette/Downloads/gotoconnect.app
                         ...  noReset=true
                         ...  bundleId=${g2_bundle_id}

**Reply by efrenette11** [reply]
The problem with that, is that when we open first we want to reinstall it.

So here's what I want to do
install and open the app
log in
close the app
reopen the app


I expect to stay log in, but there's no way to do it. Any idea ?

**Reply by KazuCocoa** [reply]
https://appium.io/docs/en/commands/device/app/terminate-app/ https://appium.io/docs/en/commands/device/app/activate-app/ may help. https://discuss.appium.io/ might be a good place to ask a question.

**Comment by jlipps** [comment]
`/appium/app/launch` is no longer supported. You're much better off doing what @KazuCocoa suggested and just using terminate app and activate app in your test script where you'd like to relaunch the app.

**Reply by efrenette11** [reply]
I'm using appium library to do so and there's no such keyword. https://serhatbolsu.github.io/robotframework-appiumlibrary/AppiumLibrary.html

Sorry for the dummy question, but how can I use those methods directly using appium library ?

**Reply by KazuCocoa** [reply]
I don't know if robotframework already has, but these apis are in python client.
https://appium.github.io/python-client-sphinx/webdriver.extensions.html?highlight=activate#webdriver.extensions.applications.Applications.activate_app
https://appium.github.io/python-client-sphinx/webdriver.extensions.html?highlight=activate#webdriver.extensions.applications.Applications.terminate_app

**Reply by efrenette11** [reply]
Tks so much guys.
